### PR TITLE
Fixes 349: show status of the build on PR

### PR DESCRIPTION
**Pull request by sch00lb0y**
## Description

I have used status api to show the status of the documentation build on each PR
## Screenshots (if appropriate):
status

## Types of changes

- [ ] Bug fix 
- [x] New feature

## Checklist:

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [x] There is a corresponding issue for this pull request.
- [x] Mentioned the Issue number in the pull request commit message `Fixes #<number> commit message`
- [x] There is only one commit per issue.

### CODE REVIEW

**Comment by imujjwal96** [comment]
Make this optional and disabled by default. We do not want to increase the load on our server just yet. 
A user may be allowed to enable this feature by going to the settings page of their repositories.

File: routes/ci.js [hunk 1]
```diff
@@ -191,11 +194,52 @@ router.post('/webhook', function(req, res, next) {
           status: true
         });
         break;
+      case "pull_request":
+        if (req.body.action === "reopened" || req.body.action ==="opened") {
+          var sha = req.body.pull_request.head.sha;
+          github.createStatus(sha, repositoryName, "pending", "Yaydoc is checking your build", process.env.PERSONAL_ACCESS_TOKEN, function (error, body) {
```
Comment (imujjwal96) [inline routes/ci.js]: What do we need this environment variable for?

File: routes/ci.js [hunk 2]
```diff
@@ -191,11 +194,52 @@ router.post('/webhook', function(req, res, next) {
...
+            if (!error) {
+              var user = req.body.pull_request.head.label.split(":")[0];
+              var targetBranch = req.body.pull_request.head.label.split(":")[1];
+              var gitURL = `https://github.com/${user}/${req.body.repository.name}.git`;
+              console.log(gitURL);
+              console.log(targetBranch);
+              var data = {
+                email: "rbalajis25@gmail.com",
```
Comment (imujjwal96) [inline routes/ci.js]: Define a non-existent email, unless you want to receive the email message for every PR :wink:

File: routes/ci.js [hunk 3]
```diff
@@ -191,11 +194,52 @@ router.post('/webhook', function(req, res, next) {
...
+            }
+          });
+        }
+        break;
       default:
         return res.json({
           status: false,
           description: 'undefined event'
-        });
+     });
```
Comment (imujjwal96) [inline routes/ci.js]: these are closing brackets for res.json. Indent them.

File: routes/ci.js [hunk 4]
```diff
@@ -191,11 +194,52 @@ router.post('/webhook', function(req, res, next) {
...
+              var data = {
+                email: "rbalajis25@gmail.com",
+                gitUrl: gitURL,
+                targetBranch: targetBranch,
+                docTheme: "",
+                debug: true,
+                docPath: "",
+                buildStatus: true
+              }
+              console.log(data);
```
Comment (imujjwal96) [inline routes/ci.js]: since it is no more a WIP, i think there is no need to console.log the data

File: backend/github.js [hunk 5]
```diff
@@ -96,3 +96,35 @@ exports.deleteHook = function (name, hook, accessToken, callback) {
     }
   });
 };
+
+/**
+ * Create status to the commit
+ * @param sha: commit sha
```
Comment (imujjwal96) [inline backend/github.js]: `commitId` makes more sense

File: backend/github.js [hunk 6]
```diff
@@ -96,3 +96,35 @@ exports.deleteHook = function (name, hook, accessToken, callback) {
     }
   });
 };
+
+/**
+ * Create status to the commit
```
Comment (imujjwal96) [inline backend/github.js]: Use a more description description :)

File: backend/github.js [hunk 7]
```diff
@@ -96,3 +96,35 @@ exports.deleteHook = function (name, hook, accessToken, callback) {
...
+ * @param sha: commit sha
+ * @param repositoryName: name of the repository. example: fossasia/yaydoc
```
Comment (imujjwal96) [inline backend/github.js]: `: 'full_name' of the repository` (remove example)

File: backend/github.js [hunk 8]
```diff
@@ -96,3 +96,35 @@ exports.deleteHook = function (name, hook, accessToken, callback) {
...
+ * @param state: state of the commit. accepted parameters are pending, success, failure, error
+ * @param description: short description of the status
+ * @param personalAccessToken: access token of the user who having access to the repository
```
Comment (imujjwal96) [inline backend/github.js]: `accessToken` as we have used everywhere else

File: backend/github.js [hunk 9]
```diff
@@ -96,3 +96,35 @@ exports.deleteHook = function (name, hook, accessToken, callback) {
...
+  request.post({
+    url: `https://api.github.com/repos/${repositoryName}/statuses/${sha}`,
+    headers: {
+      'User-Agent': 'Yaydoc',
+      'Authorization': 'token ' + personalAccessToken
+    },
+    "content-type": "application/json",
+    body: JSON.stringify({
+      state: state,
+      target_url: `https://yaydoc.herokuapp.com`,
```
Comment (imujjwal96) [inline backend/github.js]: `process.env.hostname`

File: backend/github.js [hunk 10]
```diff
@@ -96,3 +96,35 @@ exports.deleteHook = function (name, hook, accessToken, callback) {
...
+      description: description,
+      context: "Yaydoc CI"
+    })
+  }, function(error, response, body) {
+    if (error!== null) {
+      callback({description: 'Unable to create status'}, null)
```
Comment (imujjwal96) [inline backend/github.js]: `return callback().....` 
Making this change, remove the following else, resulting in 

```
if (error) {
  return callback({description: 'Unable to create status'}, null);
}
callback(null, JSON.parse(body));
```

**Comment by sch00lb0y** [comment]
@imujjwal96 updated http://demo08.herokuapp.com/

File: backend/generator.js [hunk 11]
```diff
@@ -54,7 +54,10 @@ exports.executeScript = function (socket, formData, callback) {
     var data = { code: code, email: email, uniqueId: uniqueId, gitUrl: gitUrl };
     if (code === 0) {
       data.exitCode = code;
-      if (process.env.SURGE_LOGIN === undefined || process.env.SURGE_TOKEN === undefined) {
+      if (formData.buildStatus === true) {
+        callback(null, {messages: `Process exited with code : ${code}`});
+      } else {
+             if (process.env.SURGE_LOGIN === undefined || process.env.SURGE_TOKEN === undefined) {
```
Comment (imujjwal96) [inline backend/generator.js]: check indentation

File: routes/ci.js [hunk 12]
```diff
@@ -116,15 +116,14 @@ router.post('/register', authMiddleware.isLoggedIn, function (req, res, next) {
 
 router.post('/webhook', function(req, res, next) {
   var event = req.get('X-GitHub-Event');
-  var branch = req.body.ref.split("/")[2];
```
Comment (imujjwal96) [inline routes/ci.js]: let this stay as it is.

File: routes/ci.js [hunk 13]
```diff
@@ -116,15 +116,14 @@ router.post('/register', authMiddleware.isLoggedIn, function (req, res, next) {
...
-  if (branch === "gh-pages") {
-    return res.json({
-      status: false,
-      description: "No operation on pushes to gh-pages branch"
-    });
-  }
   var query = {};
+  if (event === "pull_request") {
+    branch = req.body.pull_request.base.ref;
+  } else {
```
Comment (imujjwal96) [inline routes/ci.js]: After reverting the about change, there won't be a need to define else

Comment (sch00lb0y) [inline routes/ci.js]: @imujjwal96 I have check your suggestion. when I'm splitting undefined route is freezing.

Comment (imujjwal96) [inline routes/ci.js]: oh! right. My mistake

**Comment by imujjwal96** [comment]
Enabled pull request status from settings. Yet i cannot see yaydoc here:
image

**Comment by sch00lb0y** [comment]
@imujjwal96 I have tested in my repo it is working fine. when I send pr to your repo webhook is not triggering. I have cleared the database. please delete the webhook and check once again.

**Comment by imujjwal96** [comment]
Still not working

**Comment by sch00lb0y** [comment]
@imujjwal96 `.yaydoc.yml` is not present in your repository that's why it is not working please add the yaydoc configuration and check

**Comment by sch00lb0y** [comment]
@imujjwal96 please check this :sweat_smile:

**Comment by imujjwal96** [comment]
I hope you wouldn't mind if i review it tomorrow. Javdocs issue has left me frustrated and am exhausted for the day.

**Comment by imujjwal96** [comment]
It has been more than 8 minutes!
image
How long should I wait to get a response?

**Comment by sch00lb0y** [comment]
@imujjwal96 I have fixed the bug. Now it'll work fine. add `.yaydoc.yml` to make it work

**Comment by imujjwal96** [comment]
@sch00lb0y When i send a PR from `sam` branch to `documentation` branch, it get status code 5

**Comment by sch00lb0y** [comment]
@imujjwal96 I have fixed it now.

**Comment by imujjwal96** [comment]
What was the error?

**Comment by imujjwal96** [comment]
Can you change the url of `details` to point it to 'process.env.HOSTNAME/<reponame>' when the build is completed?

**Comment by sch00lb0y** [comment]
>What was the error?
I forgot to mention target branch when I'm resolving conflicts.
>Can you change the url of details to point it to 'process.env.HOSTNAME/' when the build is completed?
I have changed

**Comment by imujjwal96** [comment]
> I have changed

Where did you make the change? I cannot locate it! Also it is not working in my PR https://github.com/imujjwal96/prelimQuiz/pull/45

**Comment by sch00lb0y** [comment]
are you asking me to change `Process exited with code : 0 ` to url ?

**Comment by imujjwal96** [comment]
image
Details anchor tag on the right!

**Comment by sch00lb0y** [comment]
@imujjwal96 when I click detail it is redirecting to `process.env.HOSTNAME` only

**Comment by sch00lb0y** [comment]
Yeah. In another PR. I'll redirect to logs. :smile:

**Comment by imujjwal96** [comment]
ok

**Pull request merged**